src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -8,204 +8,187 @@ namespace Microsoft.OData.Client
-            if (pi != null)
-            {
-                ContentLengthMethodSetter = pi.GetSetMethod();
-            }
+            _effectiveHttpMethod = args.Method;
+            _requestUrl = args.RequestUri;
+
+            _requestMessage = new HttpRequestMessage(new HttpMethod(this.ActualMethod), _requestUrl);
 
-            pi = typeof(HttpWebRequest).GetProperty("UserAgent");
-            if (pi != null)
+            // Now set the headers.
+            foreach (var keyValue in args.Headers)
             {
-                UserAgentMethodSetter = pi.GetSetMethod();
+                this.SetHeader(keyValue.Key, keyValue.Value);
             }
         }
-#endif
 
         /// <summary>
-        /// Creates a new instance of HttpWebRequestMessage.
+        /// Returns the collection of request headers.
         /// </summary>
-        /// <param name="args">Arguments for creating the request message.</param>
-        [SuppressMessage("Microsoft.Usage", "CA2214:DoNotCallOverridableMethodsInConstructors", Justification = "SetHeader is a safe virtual method to be called from the constructor")]
-        public HttpWebRequestMessage(DataServiceClientRequestMessageArgs args)
-            : base(args.ActualMethod)
+        public override IEnumerable<KeyValuePair<string, string>> Headers
         {
-            Util.CheckArgumentNull(args, "args");
-            Debug.Assert(args.RequestUri.IsAbsoluteUri, "request uri is not absolute uri");
-            Debug.Assert(
-                args.RequestUri.Scheme.Equals("http", StringComparison.OrdinalIgnoreCase) ||
-                args.RequestUri.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase),
-                "request uri is not for HTTP");
-            this.effectiveHttpMethod = args.Method;
-            this.requestUrl = args.RequestUri;
-
-            this.httpRequest = HttpWebRequestMessage.CreateRequest(this.ActualMethod, this.Url, args);
-
-            // Now set the headers.
-            foreach (var keyValue in args.Headers)
+            get
             {
-                this.SetHeader(keyValue.Key, keyValue.Value);
+                if (_requestMessage.Content != null)
+                {
+                    return HttpHeadersToStringDictionary(_requestMessage.Headers).Concat(HttpHeadersToStringDictionary(_requestMessage.Content.Headers));
+                }
+
+                return HttpHeadersToStringDictionary(_requestMessage.Headers).Concat(_contentHeaderValueCache);
             }
         }
 
-        #region Properties
-
-        /// <summary> Returns the request url. </summary>
+        /// <summary>
+        /// Gets or sets the request url.
+        /// </summary>
         public override Uri Url
         {
             get
             {
-                return this.requestUrl;
+                return _requestUrl;
             }
-
             set
             {
-                throw new NotSupportedException();
+                _requestMessage.RequestUri = value;

```

Comment:
This property (and the Method property below) are kinda not consistent in that once I set (in this case) _requestUrl from the args constructor, all subsequent calls will always be that value, however it also allows me to set the value which goes ahead and modifies something else and if I access the value again, it is still an old value
```
var args = new DataServiceClientRequestMessageArgs { RequestUri = "x" };
var message = new HttpWebRequestMessage(args);
message.Uri == "x" <= true
message.Uri = "y";
message.Uri == "y" <= false?!!
```
Same for Method (_effectiveHttpMethod)

Reply:
resolved